src/con4m/gc.c

```diff
@@ -119,11 +98,12 @@ c4m_initialize_gc()
 
         // use c4m_gc_malloc_wrapper for hatrack's zalloc function since our
         // gc allocator always returns zeroed memory.
-        hatrack_setmallocfns(c4m_gc_malloc_wrapper,
-                             c4m_gc_malloc_wrapper,
-                             c4m_gc_realloc_wrapper,
-                             c4m_gc_free_wrapper,
-                             NULL);
+        // hatrack_setmallocfns(NULL,
+        // NULL,
+        // NULL,
+        // NULL,
+        // c4m_gc_malloc_wrapper,
+        // NULL);
```

Comment:
Yes, it was not that I was seeing any issue actively.  I knew the roots weren't being tracked, and was about to go in there and figure that out.

However, it felt much easier when using hatrack data structures inside the GC to bypass the GC (I think that's probably related to the occasional problem I had with the list of cross-heap references that was getting alloc'd but not even used). 

Hatrack would need significant refactoring to do per-datastructure memory management, and I'm not convinced it *should* be done for when there's high transaction volumes. So at the beginning I really was only assessing two options:
1. Have everything GC by default, but do the work to make sure nothing goes berzerk inside the GC since the list of roots, etc. would use the current heap as the backing store.
2. Have two sets of malloc fns, one for internal data structure allocation, and one for allocating new top-level objects.

The first was going to make things really hairy forever, so seemed like a bad choice. 

The second path seemed like the right choice; I needed to add finalizers anyway for file descriptors inside stream objects.

But doing the second properly was still a bunch of work-- I went far down the path and backed it out (the commented out code is something I missed when backing it out; note it has an extra parameter there). It was still more involved than I really was willing to do right now.

The only place where that was going to be an issue in my current use of hatrack was on set operations, which is why I ended up with the compromise of having those functions take a fully allocated result object as a parameter, instead of having the operations themselves do the alloc.

I think that's the major place where it's important anyway. My memory is that most of the other places where hatrack is allocing a toplevel object is just for _new() wrappers that aren't particularly necessary. But there might be some places like inside `logring` or in the bits I never finished like `vector`.

Anyway, I'll remove the commented code in my current branch, so that I don't need to get it reapproved :)

Reply:
I'm having trouble understanding why allocating external top-level objects should be any different than allocating internal data. Hatrack shouldn't need to know anything about roots or gc or any of that, so if an allocated external object needs to be a gc root, it's on con4m to make that so. I've been through all of the Hatrack code several times over now for various reasons, but can't think of a place where internal data would need to be a gc root other than for the `mmm_thread_t` tls data. But that can be marked as a root by specifying a custom acquisition function in con4m via `mmm_setthreadfns`, which is what I had planned to do.

I don't have any issue with making the change to not use gc for Hatrack allocations, but I think if something using Hatrack (whether that's con4m or otherwise) really needs to distinguish between external and internal allocations then it should do the external allocations itself and use the init/cleanup functions instead of the new/delete functions. Since Hatrack already supports this sort of thing, more or less, adding a separate set of allocation functions for the new/delete stuff seems unnecessary. Although I suppose there's the case of views, which are external allocations, but intended to be ephemeral.

I admittedly haven't spent the time to understand how the GC in con4m works yet, but I'll spend some time with it in the next couple of days.